leios-sim/src/Leios/Model.hs

```diff
@@ -0,0 +1,537 @@
+
+init ::
+  forall m.
+  (Monad m, MonadSTM m, Applicative m) =>
+  TVar m Parameters ->
+  m (OutsideWorld m)
+init paramsTVar = do
+  pqsTVar <- newTVarIO mempty
+  storedIBsTVar <- newTVarIO mempty
+  pure $ OutsideWorld{pqsTVar = pqsTVar, paramsTVar = paramsTVar, storedIBsTVar = storedIBsTVar}
+
+getParams :: MonadSTM m => OutsideWorld m -> m Parameters
+getParams = readTVarIO . paramsTVar
+
+-- | .
+--
+-- PRECONDITION: the node should not have been registered.
+register ::
+  forall m.
+  (Monad m, MonadSTM m) =>
+  NodeId ->
+  OutsideWorld m ->
+  m ()
+register nodeId world = do
+  pqTVar <- newPQueue
+  -- TODO: we could use broadcast channels.
+  atomically $ do
+    -- FIXME: assert the 'nodeId' is not registered.
+    modifyTVar' (pqsTVar world) (Map.insert nodeId pqTVar)
+
+-- | ...
+--
+-- In this first iteration we assume the message is broadcast to each node.
+--
+-- NOTE: a node will receive it's own message. We could prevent this by including the senders's id in the function type.
+sendTo ::
+  forall m.
+  MonadSTM m =>
+  Msg ->
+  OutsideWorld m ->
+  m ()
+sendTo msg world = atomically $ do
+  pqsMap <- readTVar (pqsTVar world)
+  for_ (Map.elems pqsMap) (insertPQueue msg)
+
+-- | ...
+--
+-- PRECONDITION: The node ID must exist.
+receiveFrom :: forall m. (MonadSTM m, MonadDelay m) => NodeId -> OutsideWorld m -> m Msg
+receiveFrom nodeId world = do
+  m_pqTVar <- fmap (Map.lookup nodeId) $ atomically $ readTVar (pqsTVar world)
+  case m_pqTVar of
+    Nothing -> error $ "Node " <> show nodeId <> " does not exist."
+    Just pqTVar -> do
+      msg <- pop pqTVar
+      owNodeBandwidth <- nodeBandwidth <$> (atomically $ readTVar (paramsTVar world))
+      let msToInt (Microseconds ms) = fromIntegral ms
+      threadDelay $ msToInt $ transmissionTime (size msg) owNodeBandwidth
+      pure msg
+
+--------------------------------------------------------------------------------
+-- Outside World Storage
+--------------------------------------------------------------------------------
+
+storeIB :: MonadSTM m => NodeId -> IB -> OutsideWorld m -> m ()
+storeIB nodeId ib OutsideWorld{storedIBsTVar} =
+  atomically $
+    modifyTVar' storedIBsTVar (Map.alter (Just . maybe [ib] (ib :)) nodeId)
+
+-- | Retrieve the downloaded IBs by the given node, which correspond to the given slice.
+storedIBsBy :: MonadSTM m => NodeId -> OutsideWorld m -> Slice -> m [IB]
+storedIBsBy nodeId OutsideWorld{storedIBsTVar} slice = do
+  mStoredIBs <- atomically (readTVar storedIBsTVar)
+  pure $
+    maybe [] (filter ((`isWithin` slice) . ib_slot)) $
+      Map.lookup nodeId mStoredIBs
+
+--------------------------------------------------------------------------------
+-- A priority queue inside a transactional var
+--------------------------------------------------------------------------------
+
+data PQueueTVar m = PQueueTVar {getTQueueTVar :: TVar m (MaxQueue PMsg)}
```

Comment:
Oh, right, I could have used [maximum](https://www.stackage.org/haddock/lts-22.26/base-4.18.2.1/src/Data.Foldable.html#maximum). However, now that we have a `PQueue`, it might make sense to leave it for efficiency's sake. I know it's a prototype, but the queue is already in place. I don't mind using either, though. Let me know 👍

Reply:
Sure, we can use `PQueue`